### PR TITLE
Automatically Finding the Centre of Rotation on a Sample

### LINKED ISSUE #17914: Imaging: Review reconstruction command script for running remote reconstruciton
Review `mantid/scripts/Imaging/IMAT/tomorec/reconstruction_command.py` for any improvements. 

The target of this review will be to determine the best way to add support for `Savu` and/or improve the way tools are handled so that future tools would be easier to add.

### LINKED ISSUE #17995: Imaging GUI: Add option to auto find center of rotation for stack of images
Add option to auto find center of rotation for stack of images

- Also fixes a bug in the ROI where the selection coordinates would be wrong depending on the direction in which the user dragged the selection area

### PR DESCRIPTION
**Pull request by dtasev**
- [x] ~Requires #18095 merged first.~

#### NOTE: This PR should only check the new functionality of `reconstruction_command.py` -> the `find_cor` function, and the functions used within. The rest of the script is currently unstable and will be fixed in #18379, as there have been too many broken things to be covered in this PR as well.

1. The reconstruction scripts -> `tomo_reconstruct.py` and `reconstruction_command.py`:
   - Added additional functions to support finding the center of rotation.
   - The new functionality can be accessed if a `-f` or `--find-cor` are passed.

2. GUI:
   - Auto COR functionality now works. 
      - It requires TomoPy installed. 
      - It can be accessed from `ROI etc.` tab, the `Auto-find` button.
      - It uses the selected `Compute Resource`, i.e. if SCARF is selected a SCARF submission will be attempted.
   - Center of Rotation is automatically read from the `StdOut` of the process
      - The way the Center of Rotation is read is quite -> [hacky](https://github.com/mantidproject/mantid/blob/54a29f24e71fa76d733f15f50643a668c4313d30/MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp#L1332), I'm open to suggestions for improvement. The last output from the find centre of rotation run is expected to _always_ be the number of the Centre of Rotation.
   - Center of Rotation is automatically loaded into the GUI
   - BUG FIX: Prompt for cancelling the reconstruction prompt only activates if the process successfully executed
   - BUG FIX: GUI would crash Mantid if the external process crashed/failed to launch, as it would still try to read the empty string and segfault.
   - Loading an image through the `ROI etc.` tab _now changes_ the paths for the sample in the `Setup` tab
   - Depending on whether a `Reconstruction` or a `Auto COR` has been run, the name of the process is now different in the `Reconstruction Jobs` on the `Run` tab

3. ImageROI is renamed to TomographyROI as I think it is clearer. More suggestions for name changes are welcome.

4. The system tests for the reconstruction scripts -> `ImagingIMATTomoScripts` are completely _disabled_ and will be rewritten in #18386. Unit tests are still present, however they will too be rewritten by the mentioned issue. 

<br>

## To test:

- Open Mantid Tomographic Reconstruction GUI
- Go on Setup tab
- Go on Run tab, select Local as computing resource, as mentioned above, it uses the selected `Compute Resource`, i.e. if SCARF is selected a SCARF submission will be attempted.
- Select a small sample (the one from #18095 will do)
- Flat and Dark paths should be detected automatically
- Go on `ROI etc.` tab and Load an image (or the whole stack)
- Check that the paths in the `Setup` tab are accurate, they should be automatically changed
- Auto-find button for Center of Rotation should now be active with only 1 option -> TomoPy

  - Windows:
    - Try running the auto find (it will likely fail)
    - See that the process was added in the Reconstruction Jobs list
    - Mantid should still run fine after this, nothing should be changed in the `ROI etc` tab (note: when the external process failed to launch mantid would crash, but that should be fixed [c58083](https://github.com/mantidproject/mantid/pull/18258/commits/c5808354f7cff6a035365592bc57c8b1c49162d7)

  - Linux:
	- Configure the Region of Interest
    - Run the auto-find tool
    - Look at the script's output, it has been changed to be very verbose and to print elapsed times on each step
      - The >> Memory is intentionally left there for now for debugging on SCARF
    - When the process is finished, see that the calculated COR is loaded into the GUI. The `+` for the COR should be placed at exactly the middle of the image and at the correct X position

  - Repeat the above step with those 2 configurations:
    1. With ROI that's not the full image
	2. With ROI that is the full image, i.e. nothing is cropped

  - Make sure the COR's are always correct to the crop, if the ROI is the full image then both the (REL for relative) REL to CROP and REL to FULL should be the same number. The code for that is [here](https://github.com/mantidproject/mantid/blob/55ca7b80eb1f219e102584fe1c1c034ac774c9cc/scripts/Imaging/IMAT/tomorec/reconstruction_command.py#L1319). 

Release Notes added to `imaging.rst`

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Tomography/TomographyCommandArguments.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#ifndef MANTIDQTCUSTOMINTERFACES_TOMOGRAPHY_TOMOGRAPHYCOMMANDARGUMENTS_H_
```
Comment (DimitarTasev) [inline MantidQt/CustomInterfaces/inc/MantidQtCustomInterfaces/Tomography/TomographyCommandArguments.h]: File not part of PR. Will be removed

**Comment by DimitarTasev** [comment]
Unit tests for the GUI and the System tests for the reconstruction scripts are bad and need to be re-designed and rewritten. That will be covered in #18386, the unit and system tests present are just changed so that they pass the build servers, and do not cause any problems to other people.

**Comment by DimitarTasev** [comment]
A unit test that sometimes crashes locally has been removed in [commit 0797b90] (https://github.com/mantidproject/mantid/pull/18258/commits/0797b905e8d250e914a42d6f5896afe4fa6f5878). This is not a problem as the whole functionality is removed in the next PR: #18134

File: MantidQt/CustomInterfaces/src/Tomography/TomographyIfacePresenter.cpp [hunk 2]
```diff
@@ -146,6 +150,75 @@ void TomographyIfacePresenter::notify(
...
+  prepareSubmissionArguments(local, runnable, args, allOpts);
+
+  // append the additional args for now
+  for (const auto &arg : cachedArgs) {
+    args.emplace_back(arg);
+    allOpts += arg;
+  }
+
+  // if no external provided, use the default one from settings
+  setupAndRunLocalExternalProcess(!cachedExec.empty() ? cachedExec : runnable,
```
Comment (DavidFair) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyIfacePresenter.cpp]: Could we split this operation out: `!cachedExec.empty() ? cachedExec : runnable` as there are several operations running within a function call meaning you have to read it very carefully

File: MantidQt/CustomInterfaces/src/Tomography/TomographyIfaceViewQtGUI.cpp [hunk 3]
```diff
@@ -177,7 +177,7 @@ void TomographyIfaceViewQtGUI::initLayout() {
   m_ui.tabMain->addTab(tabSetupW, QString("Setup"));
 
   // this is a Qt widget, let Qt manage the pointer
-  m_tabROIW = new ImageROIViewQtWidget(m_ui.tabMain);
+  m_tabROIW = new TomographyROIViewQtWidget(m_ui.tabMain);
```
Comment (DavidFair) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyIfaceViewQtGUI.cpp]: Could we rename this from `m_tabROIW` to `m_tabROIWidget` as initially I thought it was `m_tabROIWindow`?

File: MantidQt/CustomInterfaces/src/Tomography/TomographyIfaceViewQtGUI.cpp [hunk 4]
```diff
@@ -1276,14 +1285,7 @@ void TomographyIfaceViewQtGUI::flatsPathCheckStatusChanged(int status) {
 
 void TomographyIfaceViewQtGUI::darksPathCheckStatusChanged(int status) {
   bool enable = 0 != status;
```
Comment (DavidFair) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyIfaceViewQtGUI.cpp]: `bool enable (0 != status);` is much nicer to read. Additionally what does 0 represent in this case?

File: MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp [hunk 5]
```diff
@@ -1237,22 +1296,72 @@ void ImageROIViewQtWidget::updateColorMap(const std::string &filename) {
...
+  // this should run a --find-cor, empty executable path string signifies that
+  // the default external interpretor will be used
+  emit(findCORClicked("", {"--find-cor"}));
+}
+
+void TomographyROIViewQtWidget::readCoRFromProcessOutput(const QString &str) {
+  if (str.isEmpty()) {
+    // the process string is empty, it is likely the process crashed or was
+    // unable to execute properly
+    return;
```
Comment (DavidFair) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp]: Should we maybe print a warning to the log as well as I'm guessing this is pretty exceptional?

File: MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp [hunk 6]
```diff
@@ -1237,22 +1296,72 @@ void ImageROIViewQtWidget::updateColorMap(const std::string &filename) {
...
+  }
+
+  std::string output = str.toStdString();
```
Comment (DavidFair) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp]: `const`

File: MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp [hunk 7]
```diff
@@ -1237,22 +1296,72 @@ void ImageROIViewQtWidget::updateColorMap(const std::string &filename) {
...
+
+  // -- to not be on the null character
```
Comment (DavidFair) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp]: Could we add a comment saying what the expected output is meant to be so people can see why we parse it like this

File: MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp [hunk 8]
```diff
@@ -1237,22 +1296,72 @@ void ImageROIViewQtWidget::updateColorMap(const std::string &filename) {
...
+  auto back_iterator = --(output.cend());
+  std::string cor_number;
+  for (; back_iterator != output.begin(); --back_iterator) {
+    if (*back_iterator == '\n') {
+      // found the last new line
+      cor_number = std::string(back_iterator + 1, output.cend());
+      break;
+    }
+  }
+  int cor = 0;
```
Comment (DavidFair) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp]: Could we have a more descriptive name than `cor` such as `centreOfRotationValue`?

File: docs/source/release/v3.9.0/imaging.rst [hunk 9]
```diff
@@ -8,17 +8,18 @@ Imaging Changes
...
-- Local reconstruction processes are now also updated in the reconstruction jobs list
-- The reconstruction jobs list command line is now Read Only
-- Clicking Cancel now cancels the running reconstruction process
-- The reconstruction scripts will now work with TomoPy v1.x.x, however the results have yet to be tested
-- Selecting the Center of Rotation, Area for Normalsation and Region of Interest will now always follow the exact position of the mouse
+- The external interpreter and scripts paths are no longer ignored and are now correctly appended when running a local reconstruction.
+- Local reconstruction processes are now also updated in the reconstruction jobs list.
+- The reconstruction jobs list command line is now Read Only.
+- Clicking Cancel now cancels the running reconstruction process.
+- The reconstruction scripts will now run with TomoPy v1.x.x, however the output is not tested with versions newer than 0.10.x.
```
Comment (DavidFair) [inline docs/source/release/v3.9.0/imaging.rst]: Is there any way to query TomoPy for its version an provide a warning for users at runtime?

File: scripts/Imaging/IMAT/prep/filters.py [hunk 10]
```diff
@@ -33,32 +33,42 @@ def scale_down(data_vol, block_size, method='average'):
...
+
+    rescaled_vol = np.zeros(
+        (data_vol.shape[0], data_vol.shape[1] // block_size,
+         data_vol.shape[2] // block_size),
+        dtype=data_vol.dtype)
     # for block averages in every slice/image along the vertical/z axis
-    tmp_shape = rescaled_vol.shape[1], block_size, rescaled_vol.shape[2], block_size
+    tmp_shape = rescaled_vol.shape[1], block_size, rescaled_vol.shape[
+        2], block_size
     for vert_slice in range(len(rescaled_vol)):
```
Comment (DavidFair) [inline scripts/Imaging/IMAT/prep/filters.py]: Use `xrange` if rescaled_vol is large

File: scripts/Imaging/IMAT/tomo_reconstruct.py [hunk 11]
```diff
@@ -54,10 +54,10 @@
 import os
 from os import path
 # So insert in the path the directory that contains this file
-sys.path.insert(0, os.path.split(path.dirname(__file__))[0]) # noqa
+sys.path.insert(0, os.path.split(path.dirname(__file__))[0])  # noqa
```
Comment (DavidFair) [inline scripts/Imaging/IMAT/tomo_reconstruct.py]: WRGT sys.path.insert: http://stackoverflow.com/a/10097543

Additionally what are we appending to the system path - personally this line scares me, isn't this what virtual environments are for?

File: scripts/Imaging/IMAT/tomo_reconstruct.py [hunk 12]
```diff
@@ -69,99 +69,192 @@ def setup_cmd_options():
...
+        type=str,
+        help="Methods supported: 'wf' (Wavelet-Fourier)")
+
+    grp_pre.add_argument(
+        "--rotation",
+        required=False,
+        type=int,
+        help="Rotate images by 90 degrees a number of "
+        "times. The rotation is clockwise unless a negative number is given which indicates "
+        "rotation counterclocwise")
```
Comment (DavidFair) [inline scripts/Imaging/IMAT/tomo_reconstruct.py]: Typo - "counterclockwise"

File: scripts/Imaging/IMAT/tomo_reconstruct.py [hunk 13]
```diff
@@ -69,99 +69,192 @@ def setup_cmd_options():
...
+
+    grp_pre.add_argument(
+        "--scale-down",
+        required=False,
+        type=int,
+        help="Scale down factor, to reduce the size of "
+        "the images for faster (lower-resolution) reconstruction. For example a factor of 2 "
+        "reduces 1kx1k images to 512x512 images (combining blocks of 2x2 pixels into a single "
+        "pixel. The output pixels are calculated as the average of the input pixel blocks."
```
Comment (DavidFair) [inline scripts/Imaging/IMAT/tomo_reconstruct.py]: Does that mean the output resolution is scaled down or is it the same resolution based on the interpolated values?

File: scripts/Imaging/IMAT/tomorec/configs.py [hunk 14]
```diff
@@ -49,11 +49,13 @@ def __str__(self):
         mystr = "Tool: {0}\n".format(self.tool)
         mystr += "Algorithm: {0}\n".format(self.algorithm)
         if self.num_iter:
-            mystr += "Number of algorith iterations: {0}\n".format(self.num_iter)
+            mystr += "Number of algorith iterations: {0}\n".format(
```
Comment (DavidFair) [inline scripts/Imaging/IMAT/tomorec/configs.py]: Typo "algorithm"

**Review by DavidFair — CHANGES_REQUESTED** [review]
Some minor changes and feedback

File: MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp (hunk 5 above)
Comment (DimitarTasev) [inline MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp]: The error will be printed out in the mantid log, and the job will have status `Exit`

**Review by DavidFair — APPROVED** [review]

**Comment by DavidFair** [comment]
I have tested against Windows and Linux and the GUI works correctly following the testing instructions.

There was an issue running the centre of rotation with an invalid interpreter path then closing the interface crashing Mantid. This is an edge case so shouldn't affect the PR and an issue is opened here: #18532

The remaining comments on the .py scripts (typos mainly) aren't applicable as the scripts are in a completely different state as noted in the original description.

Once the final tests finish :shipit:

**Pull request merged**